src/Ast/Type/ArrayShapeNode.php

```diff
@@ -22,15 +25,27 @@ class ArrayShapeNode implements TypeNode
 	/** @var self::KIND_* */
 	public $kind;
 
+	/** @var GenericTypeNode|null */
+	public $extraItemType;
```

Comment:
Hey, I'd rather have a different design:

```php
/** @var TypeNode|null */
public $extraKeyType;

/** @var TypeNode|null */
public $extraValueType;

// ...

/**
 * @phpstan-assert-if-false !null $this->extraKeyType
 * @phpstan-assert-if-false !null $this->extraValueType
 */
public function isSealed(): bool
{
}

/**
 * @phpstan-assert-if-true !null $this->extraKeyType
 * @phpstan-assert-if-true !null $this->extraValueType
 */
public function isUnsealed(): bool
{
}
```

The GenericTypeNode here is too broad, it allows for invalid AST with types count that isn't 2.

Reply:
I mainly went for `GenericTypeNode` to keep consistency with regular array definitions (e.g. `array<string, string, string>` works fine on a doc parser level), but I'll happily make it more restrictive.

The assertions on the other hand seem slightly concerning, specifically for unsealed shapes without extra key/value types.
- Given `array{...}` 
- We can internally represent that as `array{...<array-key, mixed>}`
- And for printing / string conversions output that as `array{...}` again (also to keep backward compatibility)
- But this also means that an explicit `array{...<array-key, mixed>}` would also result in an output of `array{...}`, unless we add a distinction for "not set"
- Also, `list{...}` would then internally be represented as `list{...<mixed>}` (internally with an irrelvant `int` key type)

I don't mind doing that, but is the complexity really more suitable in the parser vs. handling it entirely on the type resolution side? (I suppose there's a fine line between the two)

Reply:
It's currently allowed. Is that breaking change ok? (see the PR that introduced it https://github.com/phpstan/phpdoc-parser/pull/169)

Note that my concerns also apply for e.g. `array{foo: string, ...}`

Reply:
Of note: Psalm allows `array{...}` & `list{...}`, and like `array{}` & `list{}`  resolves them to `array<never, never>`: https://psalm.dev/r/7840e76832

In PHPStan they're similarly all resolved to `array{}`: https://phpstan.org/r/69df8759-e426-4d5d-a8d3-83d26f973624

Reply:
Alright, I get it. As for the AST in phpdoc-parser, I guess we need to throw away my idea for the methods with `@phpstan-assert`. We'll keep the `public $sealed` which will still mean "are there `...` in the array shape?". And then we should have a new AST object which will contain both key type nad value type.

So there will be some kind of "ArrayShapeUnsealedTypes|null" property in ArrayShapeNode with `$keyType` and `$valueType`. Which will make it impossible to represent impossible states in the object structure.

The only impossible state to represent would be `unsealed=false` and `unsealedType !== null` but I'm fine to live with that.

As for the PHPStan typesystem representation, we can do the same thing we're doing to `array<X>` vs. `array<array-key, X>`.

`array{foo: X, ...}` would be represented with benevolent union type in the place of unsealed key type, and implicit mixed type in place of the value. So we'd still know what to do in `toPhpDocNode()`.

`array<foo: X, ...<array-key, mixed>` would still have the benevolent union type, but the mixed would be explicit, so we'd know what to do in `toPhpDocNode()`.

Thanks!

Reply:
> The only impossible state to represent would be sealed=true and unsealedType !== null but I'm fine to live with that.

There's also the impossible state of `$kind === 'list' && $unsealedType->keyType !== null`, but that's it.

I've made the necessary adjustments.